framework/src/main/java/org/checkerframework/common/wholeprograminference/scenelib/ASceneWrapper.java

```diff
@@ -0,0 +1,246 @@
+     * @param annosToIgnore which annotations should be ignored in which contexts
+     */
+    public void writeToJaif(
+            String jaifPath, Map<Pair<String, TypeUseLocation>, Set<String>> annosToIgnore) {
+        try {
+            AScene scene = theScene.clone();
+            removeIgnoredAnnosFromScene(scene, annosToIgnore);
+            scene.prune();
+            new File(jaifPath).delete();
+            if (!scene.isEmpty()) {
+                // Only write non-empty scenes into .jaif files.
+                IndexFileWriter.write(scene, new FileWriter(jaifPath));
+            }
+        } catch (IOException e) {
+            throw new UserError(
+                    "Problem while reading file in: "
+                            + jaifPath
+                            + ". Exception message: "
+                            + e.getMessage(),
+                    e);
+        } catch (DefException e) {
+            throw new BugInCF(e);
+        }
+    }
+
+    /**
+     * Write the scene represented by this object to a stub file.
+     *
+     * @param jaifPath a path that ends in ".jaif". The stub file will be created on the same path,
+     *     but the extension will be replaced with ".astub"
+     * @param annosToIgnore which annotations to ignore in which contexts
+     */
+    public void writeToStub(
+            String jaifPath, Map<Pair<String, TypeUseLocation>, Set<String>> annosToIgnore) {
+        removeIgnoredAnnosFromScene(theScene, annosToIgnore);
+        theScene.prune();
+        String stubPath = jaifPath.replace(".jaif", ".astub");
+        new File(stubPath).delete();
+        if (!theScene.isEmpty()) {
+            // Only write non-empty scenes into .astub files.
+            try {
+                SceneToStubWriter.write(this, new FileWriter(stubPath));
+            } catch (IOException e) {
+                throw new UserError("Problem while writing %s: %s", stubPath, e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Interact with scenelib to add a class to the scene.
+     *
+     * <p>Results are interned.
+     *
+     * @param className the binary name of the class to be added to the scene
+     * @param classSymbol the element representing the class, used for adding data to the
+     *     AClassWrapper returned by this method. If it is null, the AClassWrapper's data will not
+     *     be updated.
+     * @return an AClassWrapper representing that class
+     */
+    public AClassWrapper vivifyClass(
+            @BinaryName String className, @Nullable ClassSymbol classSymbol) {
+        AClassWrapper wrapper;
+        if (classes.containsKey(className)) {
+            wrapper = classes.get(className);
+        } else {
+            AClass aClass = theScene.classes.getVivify(className);
+            wrapper = new AClassWrapper(aClass);
+            classes.put(className, wrapper);
+        }
+        if (classSymbol != null) {
```

Comment:
Why does this update need to be performed on both branches rather than just the `else` branch?

I expect a method that uses a cache to return the cached value, or to create a new value.  Here, the code side-effects a cached value.

Reply:
Thanks for that information.  I understand now.  If you could add a brief version of it to the documentation, that would clarify it for future readers of the code.